src/evaluators/UnaryExpression.js

```diff
@@ -97,22 +101,33 @@ function evaluateDeleteOperation(expr: Value | Reference, realm: Realm) {
 
 function generateRuntimeCall(ref: Reference | Value, ast: BabelNodeUnaryExpression, strictCode: boolean, realm: Realm) {
   invariant(ref instanceof Reference);
-  let baseValue = Environment.GetBase(realm, ref);
-  invariant(baseValue instanceof Value);
-  let propertyName = Environment.GetReferencedName(realm, ref);
-  invariant(typeof propertyName === "string");
+  let value;
+  let propertyName;
 
-  if (!baseValue.isSimpleObject()) {
-    Leak.leakValue(realm, baseValue, ast.loc);
+  // the base reference might be an EnvironmentRecord or a ObjectValue
+  // if it is a EnvironmentRecord, we don't serialize a member expression below
+  if (Environment.IsPropertyReference(realm, ref)) {
+    value = Environment.GetBase(realm, ref);
+    invariant(value instanceof Value);
+    if (!value.isSimpleObject()) {
+      Leak.leakValue(realm, value, ast.loc);
+    }
+    propertyName = Environment.GetReferencedName(realm, ref);
+  } else {
+    value = Environment.GetValue(realm, ref);
+    invariant(value instanceof Value);
+    Leak.leakValue(realm, value, ast.loc);
   }
-  return AbstractValue.createTemporalFromBuildFunction(realm, Value, [baseValue], nodes => {
+
+  return AbstractValue.createTemporalFromBuildFunction(realm, Value, [value], nodes => {
```

Comment:
It is not obvious to me that this needs to be a temporal value. Can you add a comment to explain why it must be?

Reply:
I saw that we use `createFromUnaryOp`, but I couldn't get it working properly with `delete` operations on object properties. Rather than over-complicate the logic and create either `createFromUnaryOp` or `createTemporalFromBuildFunction` depending on the operation, I just unified them both.

Reply:
The delete operator is misclassified as a unary operator. It should really be an update operator. The way it is currently implement for abstract values appear to be broken. You use of a temporal value masks this problem in your test cases, but does not solve the underlying bug. The best way to proceed would be to do the right thing here, but file an issue to fix delete.